File: backend/src/main/java/com/matchup/auth/service/AuthService.java
package com.matchup.auth.service;

import com.matchup.auth.dto.*;
import com.matchup.auth.entity.PasswordResetToken;
import com.matchup.auth.entity.RefreshToken;
import com.matchup.auth.repository.PasswordResetTokenRepository;
import com.matchup.auth.repository.RefreshTokenRepository;
import com.matchup.common.exception.ConflictException;
import com.matchup.common.exception.InvalidTokenException;
import com.matchup.common.exception.NotFoundException;
import com.matchup.security.JwtService;
import com.matchup.user.entity.RolUsuario;
import com.matchup.user.entity.Usuario;
import com.matchup.user.repository.UsuarioRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.userdetails.User;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.time.temporal.ChronoUnit;
import java.util.UUID;

@Slf4j
@Service
@RequiredArgsConstructor
public class AuthService {

    private final UsuarioRepository usuarioRepository;
    private final RefreshTokenRepository refreshTokenRepository;
    private final PasswordResetTokenRepository passwordResetTokenRepository;
    private final JwtService jwtService;
    private final PasswordEncoder passwordEncoder;
    private final AuthenticationManager authenticationManager;

    @Value("${app.jwt.refresh-expiration-ms}")
    private long refreshExpirationMs;

    @Transactional
    public AuthResponse register(RegisterRequest request) {
        if (usuarioRepository.existsByEmail(request.getEmail())) {
            throw new ConflictException("El email ya está registrado: " + request.getEmail());
        }
        Usuario usuario = Usuario.builder()
                .nombre(request.getNombre())
                .email(request.getEmail())
                .passwordHash(passwordEncoder.encode(request.getPassword()))
                .rol(RolUsuario.USER)
                .build();
        usuarioRepository.save(usuario);

        return buildAuthResponse(usuario);
    }

    @Transactional
    public AuthResponse login(LoginRequest request) {
        authenticationManager.authenticate(
                new UsernamePasswordAuthenticationToken(request.getEmail(), request.getPassword())
        );
        Usuario usuario = usuarioRepository.findByEmail(request.getEmail())
                .orElseThrow(() -> new NotFoundException("Usuario no encontrado"));

        return buildAuthResponse(usuario);
    }

    @Transactional
    public AuthResponse refresh(RefreshTokenRequest request) {
        RefreshToken stored = refreshTokenRepository.findByToken(request.getRefreshToken())
                .orElseThrow(() -> new InvalidTokenException("Refresh token no válido"));

        if (stored.getRevoked() || stored.getExpiryDate().isBefore(LocalDateTime.now())) {
            throw new InvalidTokenException("Refresh token expirado o revocado");
        }

        // Rotar: revocar el antiguo y emitir uno nuevo
        stored.setRevoked(true);
        refreshTokenRepository.save(stored);

        return buildAuthResponse(stored.getUsuario());
    }

    @Transactional
    public void logout(LogoutRequest request) {
        refreshTokenRepository.findByToken(request.getRefreshToken()).ifPresent(rt -> {
            rt.setRevoked(true);
            refreshTokenRepository.save(rt);
        });
    }

    @Transactional
    public void forgotPassword(ForgotPasswordRequest request) {
        // Silencioso si el email no existe (evita user enumeration)
        usuarioRepository.findByEmail(request.getEmail()).ifPresent(usuario -> {
            String tokenValue = UUID.randomUUID().toString();
            PasswordResetToken prt = PasswordResetToken.builder()
                    .token(tokenValue)
                    .usuario(usuario)
                    .expiryDate(LocalDateTime.now().plusHours(1))
                    .build();
            passwordResetTokenRepository.save(prt);
            // TODO: reemplazar por envío real de email cuando se configure Spring Mail
            log.info("[DEV] Password reset token para {}: {}", request.getEmail(), tokenValue);
        });
    }

    @Transactional
    public void resetPassword(ResetPasswordRequest request) {
        PasswordResetToken prt = passwordResetTokenRepository.findByToken(request.getToken())
                .orElseThrow(() -> new InvalidTokenException("Token de reset no válido"));

        if (prt.getUsed() || prt.getExpiryDate().isBefore(LocalDateTime.now())) {
            throw new InvalidTokenException("Token de reset expirado o ya utilizado");
        }

        Usuario usuario = prt.getUsuario();
        usuario.setPasswordHash(passwordEncoder.encode(request.getNewPassword()));
        usuarioRepository.save(usuario);

        prt.setUsed(true);
        passwordResetTokenRepository.save(prt);
    }

    // ── helpers ──

    private AuthResponse buildAuthResponse(Usuario usuario) {
        UserDetails userDetails = toUserDetails(usuario);
        String accessToken = jwtService.generateAccessToken(userDetails);
        String refreshToken = issueRefreshToken(usuario);
        return AuthResponse.builder()
                .accessToken(accessToken)
                .refreshToken(refreshToken)
                .build();
    }

    private String issueRefreshToken(Usuario usuario) {
        String tokenValue = UUID.randomUUID().toString();
        RefreshToken rt = RefreshToken.builder()
                .token(tokenValue)
                .usuario(usuario)
                .expiryDate(LocalDateTime.now().plus(refreshExpirationMs, ChronoUnit.MILLIS))
                .build();
        refreshTokenRepository.save(rt);
        return tokenValue;
    }

    private UserDetails toUserDetails(Usuario usuario) {
        return User.builder()
                .username(usuario.getEmail())
                .password(usuario.getPasswordHash())
                .roles(usuario.getRol().name())
                .build();
    }
}

File: frontend/src/app/core/services/auth.service.ts
import { Injectable, signal, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Router } from '@angular/router';
import { Observable, tap } from 'rxjs';
import { environment } from '../../../environments/environment';
import { TokenService } from './token.service';
import {
  AuthResponse,
  ForgotPasswordRequest,
  LoginRequest,
  RegisterRequest,
  ResetPasswordRequest,
} from '../models/auth.model';
import { UsuarioResponse } from '../models/user.model';

@Injectable({ providedIn: 'root' })
export class AuthService {
  private readonly http = inject(HttpClient);
  private readonly tokenService = inject(TokenService);
  private readonly router = inject(Router);

  private readonly base = `${environment.apiUrl}/auth`;

  readonly currentUser = signal<UsuarioResponse | null>(null);

  register(body: RegisterRequest): Observable<AuthResponse> {
    return this.http.post<AuthResponse>(`${this.base}/register`, body).pipe(
      tap(res => this.tokenService.setTokens(res.access_token, res.refresh_token))
    );
  }

  login(body: LoginRequest): Observable<AuthResponse> {
    return this.http.post<AuthResponse>(`${this.base}/login`, body).pipe(
      tap(res => this.tokenService.setTokens(res.access_token, res.refresh_token))
    );
  }

  refresh(): Observable<AuthResponse> {
    const refreshToken = this.tokenService.getRefreshToken();
    return this.http
      .post<AuthResponse>(`${this.base}/refresh`, { refreshToken })
      .pipe(tap(res => this.tokenService.setTokens(res.access_token, res.refresh_token)));
  }

  logout(): void {
    const refreshToken = this.tokenService.getRefreshToken();
    if (refreshToken) {
      this.http.post(`${this.base}/logout`, { refreshToken }).subscribe({ error: () => {} });
    }
    this.tokenService.clear();
    this.currentUser.set(null);
    this.router.navigate(['/auth/login']);
  }

  forgotPassword(body: ForgotPasswordRequest): Observable<unknown> {
    return this.http.post(`${this.base}/forgot-password`, body);
  }

  resetPassword(body: ResetPasswordRequest): Observable<unknown> {
    return this.http.post(`${this.base}/reset-password`, body);
  }

  loadCurrentUser(): Observable<UsuarioResponse> {
    return this.http
      .get<UsuarioResponse>(`${environment.apiUrl}/users/me`)
      .pipe(tap(user => this.currentUser.set(user)));
  }
}
